{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7aad8d8b",
   "metadata": {},
   "source": [
    "## janesick_nature_comms_2023"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16d83caa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "from collections import namedtuple, Counter\n",
    "\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "\n",
    "## Image manipulation and geometry\n",
    "from cv2 import perspectiveTransform\n",
    "from ome_types import from_tiff\n",
    "from skimage.measure import points_in_poly\n",
    "\n",
    "from scipy.spatial import KDTree, ConvexHull\n",
    "from scipy.sparse import csc_matrix\n",
    "from shapely.geometry import LinearRing\n",
    "\n",
    "## Plotting imports\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.patches import Polygon\n",
    "from matplotlib.collections import PatchCollection\n",
    "from matplotlib.colors import LinearSegmentedColormap, to_hex, Colormap\n",
    "from matplotlib.axes import Axes\n",
    "\n",
    "\n",
    "def hexlist_to_cmap(name: str, hex_list: list[str]) -> Colormap:\n",
    "    \"\"\"\n",
    "    Given a list of hex colors produces a matplotlib discrete colormap.\n",
    "\n",
    "    \"\"\"\n",
    "    colors = [to_hex(hex_color) for hex_color in hex_list]\n",
    "    cmap = LinearSegmentedColormap.from_list(name=name, colors=colors, N=len(colors))\n",
    "    return cmap\n",
    "\n",
    "\n",
    "def polygons_coords_to_patch_collection(\n",
    "    poly_coords: list[list[float]], **patch_collection_kwargs\n",
    ") -> PatchCollection:\n",
    "    \"\"\"\n",
    "    Generates a matplotlib PatchCollection from a set of polygons. A helper function for plot_polygons.\n",
    "    \"\"\"\n",
    "    patches = [Polygon(hexagon) for hexagon in poly_coords]\n",
    "    collection = PatchCollection(patches, **patch_collection_kwargs)\n",
    "    return collection\n",
    "\n",
    "\n",
    "def plot_polygons(\n",
    "    poly_coords: list[list[float]],\n",
    "    ax: Axes = None,\n",
    "    auto_set_lim: bool = True,\n",
    "    **patch_collection_kwargs\n",
    "):\n",
    "    \"\"\"\n",
    "    Plots a set of polygons. Polygons may be styled using patch_collection_kwargs passed to Collection.\n",
    "    e.g.\n",
    "    array, facecolor, edgecolor, linewidth....\n",
    "    https://matplotlib.org/stable/api/collections_api.html#matplotlib.collections.Collection\n",
    "\n",
    "    pitfall:\n",
    "    when auto_set_lim == False, a plot will no rescale to include polygons by default.\n",
    "    \"\"\"\n",
    "    if ax is None:\n",
    "        ax = plt.gca()\n",
    "\n",
    "    ax.add_collection(\n",
    "        polygons_coords_to_patch_collection(poly_coords, **patch_collection_kwargs)\n",
    "    )\n",
    "    if auto_set_lim:\n",
    "        ax.autoscale_view()\n",
    "\n",
    "\n",
    "def hex_corner_offset(corner: int, size: float) -> np.ndarray[float, (2,)]:\n",
    "    \"\"\"\n",
    "    Helper function for polygon_corners, calculates the offset of each hexagon vertex from the center.\n",
    "    \"\"\"\n",
    "    angle_deg = 60 * corner - 30\n",
    "    angle_rad = math.pi / 180 * angle_deg\n",
    "    offset = np.array([size[0] * math.cos(angle_rad), size[1] * math.sin(angle_rad)])\n",
    "    return offset\n",
    "\n",
    "\n",
    "def polygon_corners(\n",
    "    center: np.ndarray[float, (2,)], size: float\n",
    ") -> np.ndarray[float, (6, 2)]:\n",
    "    \"\"\"\n",
    "    Generates the polygon vertices for a hexegon of size at origin center.\n",
    "    Used for plotting Visium space filling Hexagons.\n",
    "    Reference for working with Hexagons: https://www.redblobgames.com/grids/hexagons/\n",
    "    Visium spots are organized such that space filling hexagons are POINTY.\n",
    "    \"\"\"\n",
    "    return np.array([center + hex_corner_offset(i, size) for i in range(0, 6)])\n",
    "\n",
    "\n",
    "## The hexcodes for the colors used within the paper.\n",
    "## The celltypes are mapped to the colors for plotting convenience.\n",
    "celltypes, hex_codes = np.array(\n",
    "    sorted(\n",
    "        [\n",
    "            [\n",
    "                \"DCIS_1\",\n",
    "                \"#fb34cd\",\n",
    "            ],\n",
    "            [\n",
    "                \"DCIS_2\",\n",
    "                \"#FE664D\",\n",
    "            ],\n",
    "            [\n",
    "                \"Myoepi_ACTA2+\",\n",
    "                \"#009203\",\n",
    "            ],\n",
    "            [\n",
    "                \"Myoepi_KRT15+\",\n",
    "                \"#66c102\",\n",
    "            ],\n",
    "            [\n",
    "                \"Invasive_Tumor\",\n",
    "                \"#ff002a\",\n",
    "            ],\n",
    "            [\n",
    "                \"Prolif_Invasive_Tumor\",\n",
    "                \"#8e0119\",\n",
    "            ],\n",
    "            [\n",
    "                \"T_Cell_&_Tumor_Hybrid\",\n",
    "                \"#CB6035\",\n",
    "            ],\n",
    "            [\n",
    "                \"Stromal\",\n",
    "                \"#e5e022\",\n",
    "            ],\n",
    "            [\n",
    "                \"Stromal_&_T_Cell_Hybrid\",\n",
    "                \"#C1A029\",\n",
    "            ],\n",
    "            [\n",
    "                \"CD4+_T_Cells\",\n",
    "                \"#4fa9ff\",\n",
    "            ],\n",
    "            [\n",
    "                \"CD8+_T_Cells\",\n",
    "                \"#1068be\",\n",
    "            ],\n",
    "            [\n",
    "                \"B_Cells\",\n",
    "                \"#565DFD\",\n",
    "            ],\n",
    "            [\n",
    "                \"Macrophages_1\",\n",
    "                \"#10686f\",\n",
    "            ],\n",
    "            [\n",
    "                \"Macrophages_2\",\n",
    "                \"#3694a8\",\n",
    "            ],\n",
    "            [\n",
    "                \"IRF7+_DCs\",\n",
    "                \"#9f50f9\",\n",
    "            ],\n",
    "            [\n",
    "                \"LAMP3+_DCs\",\n",
    "                \"#AB76AE\",\n",
    "            ],\n",
    "            [\n",
    "                \"Mast_Cells\",\n",
    "                \"#999999\",\n",
    "            ],\n",
    "            [\n",
    "                \"Perivascular-Like\",\n",
    "                \"#515151\",\n",
    "            ],\n",
    "            [\"Endothelial\", \"#01257b\"],\n",
    "            [\"Unlabeled\", \"#ffa5aa\"],\n",
    "        ],\n",
    "        key=lambda x: x[0],\n",
    "    )\n",
    ").T\n",
    "## used for converting list of celltypes to categorical.\n",
    "ctype_to_code_map = {ctype: i for i, ctype in enumerate(celltypes)}\n",
    "## used for converting list of celltypes to colors.\n",
    "ctype_hex_map = dict(zip(celltypes, hex_codes))\n",
    "## matplotlib colormap\n",
    "ctype_cmap = hexlist_to_cmap(\"celltype_cmap\", hex_codes)\n",
    "\n",
    "\n",
    "def unique_encode(\n",
    "    array: np.ndarray[str], return_inverse: bool = True\n",
    ") -> (np.ndarray[str], np.ndarray[int]):\n",
    "    \"\"\"np.unique is slower then list(set(array)) thus this is an emulator function of np.unique.\n",
    "    for fast encoding of categorical arrays.\n",
    "    \"\"\"\n",
    "    unique_values = np.array(list(set(array)))\n",
    "    if not return_inverse:\n",
    "        return unique_values\n",
    "    inverse_map = {val: i for i, val in enumerate(unique_values)}\n",
    "    inverse_codes = np.array([inverse_map[i] for i in array])\n",
    "    return (unique_values, inverse_codes)\n",
    "\n",
    "\n",
    "def get_xenium_to_morphology_transform_from_xenium_morphology_image(\n",
    "    morphology_path: str, pyramid_level: int = 0\n",
    ") -> np.ndarray[np.float32]:\n",
    "    \"\"\"\n",
    "    The Xenium coordinate system has its origin at (0,0) set to the top left of the Xenium morphology image.\n",
    "    The image coordinate system is in uM and must be converted to px. This produces a transformation matrix for\n",
    "    that conversion.\n",
    "\n",
    "    https://kb.10xgenomics.com/hc/en-us/articles/11636252598925-What-are-the-Xenium-image-scale-factors-\n",
    "    \"\"\"\n",
    "    morphology_metadata = from_tiff(morphology_path)\n",
    "    origin_x = morphology_metadata.plates[0].well_origin_x\n",
    "    origin_y = morphology_metadata.plates[0].well_origin_y\n",
    "    physical_size_x = morphology_metadata.images[0].pixels.physical_size_x\n",
    "    physical_size_y = morphology_metadata.images[0].pixels.physical_size_y\n",
    "    ##Homography matrix describing scaling by pixel-size and translating by offset.\n",
    "    return np.array(\n",
    "        [\n",
    "            [1 / physical_size_x / 2**pyramid_level, 0, 0],\n",
    "            [0, 1 / physical_size_y / 2**pyramid_level, 0],\n",
    "            [0, 0, 1],\n",
    "        ]\n",
    "    )\n",
    "\n",
    "\n",
    "def get_xenium_capture_polygon_um(\n",
    "    path_to_xenium_bundle: str,\n",
    ") -> np.ndarray[int, (4, 2)]:\n",
    "    \"\"\"Produces a polygon which describes the Xenium morphology image shape.\n",
    "    Polygons are easily transformable. Thus, this polygon is useful for sub-setting coordinates\n",
    "    that lie within an image under arbitrary transformation, and of course for debug purposes.\n",
    "    \"\"\"\n",
    "    morphology_metadata = from_tiff(path_to_xenium_bundle / \"morphology_mip.ome.tif\")\n",
    "    image_shape_px = np.array(\n",
    "        [\n",
    "            morphology_metadata.images[0].pixels.size_x,\n",
    "            morphology_metadata.images[0].pixels.size_y,\n",
    "        ]\n",
    "    )\n",
    "    image_shape_um = (\n",
    "        image_shape_px * morphology_metadata.images[0].pixels.physical_size_x\n",
    "    )\n",
    "    xenium_capture_polygon_um = np.array(\n",
    "        [(0, 0), (0, image_shape_um[1]), image_shape_um, (image_shape_um[0], 0)]\n",
    "    )\n",
    "    return xenium_capture_polygon_um\n",
    "\n",
    "\n",
    "def transform_coordinates(\n",
    "    coords: np.ndarray[np.float32, (None, 2)], transform_matrix: np.ndarray[np.float32]\n",
    ") -> np.ndarray[np.float32, (None, 2)]:\n",
    "    \"\"\"Transforms coordinates using transform.\n",
    "    https://en.wikipedia.org/wiki/Homography_(computer_vision)\n",
    "    \"\"\"\n",
    "    return perspectiveTransform(\n",
    "        coords.reshape(-1, 1, 2).astype(np.float32), transform_matrix\n",
    "    )[:, 0, :]\n",
    "\n",
    "\n",
    "def get_median_spot_to_spot_distance_from_centroids(\n",
    "    centroids: np.ndarray[np.float32, (None, 2)]\n",
    ") -> float:\n",
    "    \"\"\"\n",
    "    Calculates the median distance between all coordinates and their closest neighbor.\n",
    "    Usefull for quickly determining the median spot <-> spot distance for Visium spots in the current coordinate system.\n",
    "    Can also be directly calculated using the full resolution image pixel size and the theoretical spot <-> spot distance.\n",
    "    This distance is used to generate a rough estimate of the Visium capture area and to produce space filling\n",
    "    visium spot Hexagons.\n",
    "    \"\"\"\n",
    "    tree = KDTree(centroids)\n",
    "    distances, neighbor = tree.query(centroids, k=2)\n",
    "    return np.median(distances[0, 1])\n",
    "\n",
    "\n",
    "def generate_space_filling_visium_polygons(\n",
    "    spot_coords: np.ndarray[np.float32, (None, 2)], spot_to_spot_distance: float = None\n",
    ") -> list[np.ndarray[float, (6, 2)]]:\n",
    "    \"\"\"\n",
    "    Generates hexagon polygons for every visium spot, by default the polygons are fully space filling.\n",
    "    That is to say the size of the hexagon is set using the spot <-> spot distance.\n",
    "    These polygons are for display purposes only and should not be used for direct analysis.\n",
    "    \"\"\"\n",
    "    if spot_to_spot_distance is None:\n",
    "        spot_to_spot_distance = get_median_spot_to_spot_distance_from_centroids(\n",
    "            spot_coords\n",
    "        )\n",
    "    size = spot_to_spot_distance / math.sqrt(3)\n",
    "    visium_hexegon_polygons = np.array(\n",
    "        [polygon_corners(coord, [size, size]) for coord in spot_coords]\n",
    "    )\n",
    "    return visium_hexegon_polygons\n",
    "\n",
    "\n",
    "def get_visium_capture_polygon(\n",
    "    spot_centroids: np.ndarray[np.float32, (None, 2)], spot_diameter: float\n",
    ") -> np.ndarray[np.float32, (None, 2)]:\n",
    "    \"\"\"\n",
    "    Generates a polygon representing the visium capture area; It calculates the convex hull of all Visium spots\n",
    "    then pads the convex hall by the diameter of one spot. This is used for filtering data outside the capture area\n",
    "    as well as visualization.\n",
    "\n",
    "    https://en.wikipedia.org/wiki/Convex_hull\n",
    "    \"\"\"\n",
    "    convex_hull = ConvexHull(spot_centroids)\n",
    "    convex_poly = convex_hull.points[convex_hull.vertices]\n",
    "\n",
    "    padded_convex_polly = np.array(\n",
    "        LinearRing(convex_poly).buffer(spot_diameter).exterior.coords.xy\n",
    "    ).T\n",
    "    return padded_convex_polly\n",
    "\n",
    "\n",
    "__OUTSIDE_VISIUM_CAPTURE_AREA_BARCODE__ = \"Outside_Visium_Capture_Area\"\n",
    "\n",
    "\n",
    "def bin_xenium_data_to_visium_spots(\n",
    "    visium_spot_centroids: np.ndarray[np.float32, (None, 2)],\n",
    "    visium_spot_diameter: float,\n",
    "    visium_barcodes: np.ndarray[str, (None, 2)],\n",
    "    xenium_coords: np.ndarray[np.float32, (None, 2)],\n",
    "    xenium_to_visium_transform: np.ndarray[np.float32],\n",
    ") -> np.ndarray[str]:\n",
    "    \"\"\"\n",
    "    Assigns xenium coordinates to visium spots using a transform into the visium coordinate system.\n",
    "    Coordinates are assigned by nearest neighbor.\n",
    "    e.g. each Xenium transcript/cell etc is assigned to the closest Visium barcode after transformation.\n",
    "    Also filters coordinates that lie outside an estimated capture area.\n",
    "    \"\"\"\n",
    "    xenium_coords_in_visium = transform_coordinates(\n",
    "        xenium_coords, xenium_to_visium_transform\n",
    "    )\n",
    "    kdtree = KDTree(visium_spot_centroids)\n",
    "    _, neighbors = kdtree.query(xenium_coords_in_visium, k=1)\n",
    "    barcode_assignments = visium_barcodes[neighbors]\n",
    "\n",
    "    visium_capture_polygon_full_res = get_visium_capture_polygon(\n",
    "        spot_centroids=visium_spot_centroids, spot_diameter=visium_spot_diameter\n",
    "    )\n",
    "    barcode_assignments[\n",
    "        ~points_in_poly(xenium_coords_in_visium, visium_capture_polygon_full_res)\n",
    "    ] = __OUTSIDE_VISIUM_CAPTURE_AREA_BARCODE__\n",
    "    return barcode_assignments\n",
    "\n",
    "\n",
    "def generate_anndata_matrix_from_transcript_assignments(\n",
    "    barcodes: np.ndarray[str],\n",
    "    feature_names: np.ndarray[str],\n",
    ") -> sc.AnnData:\n",
    "    \"\"\"Generates a count matrix as an anndata object from a set of barcodes and feature names.\n",
    "    e.g. given transcript cell assignments and gene names it will produce a barcode feature matrix.\n",
    "    \"\"\"\n",
    "    valid_feature_names, feature_codes = unique_encode(\n",
    "        feature_names, return_inverse=True\n",
    "    )\n",
    "    valid_barcodes, barcode_codes = unique_encode(barcodes, return_inverse=True)\n",
    "\n",
    "    mat_dict = Counter(zip(barcode_codes, feature_codes))\n",
    "    row_ind_col_ind = np.array(list(mat_dict.keys()), dtype=int).T\n",
    "    data = np.array(list(mat_dict.values()), dtype=np.float32)\n",
    "\n",
    "    obs = pd.DataFrame(index=valid_barcodes)\n",
    "    var = pd.DataFrame(index=valid_feature_names)\n",
    "    adata = sc.AnnData(X=csc_matrix((data, row_ind_col_ind)), obs=obs, var=var)\n",
    "    return adata"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
